src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HealthHeartbeatProvider.cs

```diff
@@ -0,0 +1,226 @@
+                throw new ArgumentNullException(nameof(payloadProvider), "Name member of IHealthHeartbeatPayloadExtension must be set to a unique value and cannot be empty");
+            }
+
+            this.payloadItems.Add(payloadProvider.Name, payloadProvider);
+        }
+
+        public void UnregisterHeartbeatPayload(string providerName)
+        {
+            if (this.payloadItems != null)
+            {
+                this.payloadItems.Remove(providerName);
+            }
+        }
+
+        #region IDisposable Support
+
+        // Override the finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
+        // ~HealthHeartbeatProvider() {
+        //   // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+        //   Dispose(false);
+        // }
+
+        // This code added to correctly implement the disposable pattern.
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            this.Dispose(true);
+            // uncomment the following line if the finalizer is overridden above.
+            // GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                }
+
+                // free unmanaged resources (unmanaged objects) and override a finalizer below.
+                // set large fields to null.
+
+                this.disposedValue = true;
+            }
+        }
+
+        #endregion
+
+        protected void Send()
+        {
+            this.SendHealthHeartbeat();
+        }
+
+        protected MetricTelemetry GatherData()
+        {
+            return this.GatherDataForHeartbeatPayload();
+        }
+
+        protected void AddDefaultPayloadItems(IDictionary<string, IHealthHeartbeatPayloadExtension> heartbeatPayloadItems)
+        {
+            try
+            {
+                heartbeatPayloadItems[string.Empty] = new HealthHeartbeatDefaultPayload(this.enabledHeartbeatPayloadFields);
+            }
+            catch (Exception e)
+            {
+                throw new ArgumentException("HealthHeartbeatProvider::AddDefaultPayloadItems : Unable to add default payload items to health heartbeat", nameof(heartbeatPayloadItems), e);
+            }
+        }
+
+        private MetricTelemetry GatherDataForHeartbeatPayload()
+        {
+            var heartbeat = new MetricTelemetry(HeartbeatSyntheticMetricName, 0.0);
+
+            foreach (var payloadItem in this.payloadItems)
+            {
+                var props = payloadItem.Value.GetPayloadProperties();
+                foreach (var kvp in props)
+                {
+                    heartbeat.Properties.Add(kvp.Key, kvp.Value.ToString());
+                }
+
+                heartbeat.Sum += payloadItem.Value.CurrentUnhealthyCount;
```

Comment:
Oh, I see. You want to query state from providers, not to let them give it to you on initialization and state change.... Some doc with example of use may be helpful.

Reply:
No, this was my original idea only. I agree with you that this leads to some dicey situations in the SDK and perhaps we can do better. Instead of actively polling (incurring cost with each interval, and that cost being determined by outside parties that may do intensive things) we should accept state updates driven by the consumer of the SDK instead, putting the control over performance on them instead of on the SDK itself.

This was exactly the type of issue I wanted to catch by putting out my PR a bit early! Thanks Sergey!